vowpalwabbit/parse_regressor.cc

```diff
@@ -262,16 +262,16 @@ void save_load_header(
 
           if (!read)
           {
-            memcpy(pair, all.pairs[i].c_str(), 2);
-            msg << all.pairs[i] << " ";
+            memcpy(pair, all.pairs[i].data(), 2);
+            msg << all.pairs[i].data() << " ";
```

Comment:
Are you sure this will do what you want it to do? Is this vector null terminated?

Its dangerous to treat a container's data() section as a null-terminated string. Please don't do that

Reply:
Would creating a temporary variable to store as a string be ok here, then?

Reply:
It would probably be ok this the same way as later in the code; using `stringstream.write()` to write a set number of bytes into the stream.

The biggest danger to that method is the fact that its possible to write nulls into the stringstream, which would cause unintuitive behavior when converting it to a string. I would say add comments to that effect and call `write()`.

nit: Additionally for futureproofing, we should memcpy `sizeof(type)*2` rather than `2`

Reply:
This would require casting right? Because write would take `char *` and we would have `unsigned char *`

Reply:
Yes, unfortunately there's no way to get around casting here since we're treating this as a printable string here. That should be fine since its only being used for debug purposes.

Reply:
Got it. Updated